panic in tracing

```
06:45:16 [ERROR] panic: PanicInfo { payload: Any { .. }, message: Some(called `Result::unwrap()` on an `Err` value: Utf8Error { valid_up_to: 17, error_len: Some(1) }), location: Location { file: "/Users/brew/Library/Caches/Homebrew/cargo_cache/registry/src/github.com-1ecc6299db9ec823/git2-0.13.21/src/tracing.rs", line: 79, col: 58 } }
trace:
   0: backtrace::capture::Backtrace::new
   1: gitui::set_panic_handlers::{{closure}}
   2: std::panicking::rust_panic_with_hook
   3: std::panicking::begin_panic_handler::{{closure}}
   4: std::sys_common::backtrace::__rust_end_short_backtrace
   5: _rust_begin_unwind
   6: core::panicking::panic_fmt
   7: core::result::unwrap_failed
   8: core::result::Result<T,E>::unwrap
   9: git2::tracing::tracing_cb_c
  10: _git_trace__write_fmt
  11: _stream_write
  12: _git_http_client_send_body
  13: _http_stream_write
  14: _stream_thunk
  15: _write_one
  16: _write_pack
  17: _git_smart__push
  18: _git_push_finish
  19: _git_remote_upload
  20: _git_remote_push
  21: git2::remote::Remote::push
  22: std::sys_common::backtrace::__rust_begin_short_backtrace
  23: core::ops::function::FnOnce::call_once{{vtable.shim}}
  24: std::sys::unix::thread::Thread::new::thread_start
  25: _pthread_jit_write_protect_np
```
this was reported here: https://github.com/extrawurst/gitui/issues/881#issuecomment-974563786

probably caused because our tracing implementation (the callback) did not expect libgit2 sending non-utf8 tracing. turns out it factually traces arbitrary buffers. so its no wonder out `to_str.unwrap` chokes on it 🙈

--- a/src/tracing.rs
+++ b/src/tracing.rs
@@ -57,6 +57,12 @@ impl Binding for TraceLevel {
     }
 }
 
+/// Callback type used to pass tracing events to the subscriber.
+/// see `trace_set` to register a scubscriber.
+///
+/// Note:
+///     libgit2 might pass non-utf8 strings therefore we
+///     pass the message as a byte slice
 pub type TracingCb = fn(TraceLevel, &[u8]);
 
 static CALLBACK: AtomicUsize = AtomicUsize::new(0);

--- a/src/tracing.rs
+++ b/src/tracing.rs
@@ -76,7 +76,7 @@ extern "C" fn tracing_cb_c(level: raw::git_trace_level_t, msg: *const c_char) {
     let cb = CALLBACK.load(Ordering::SeqCst);
     panic::wrap(|| unsafe {
         let cb: TracingCb = std::mem::transmute(cb);
-        let msg = std::ffi::CStr::from_ptr(msg).to_str().unwrap();
-        cb(Binding::from_raw(level), msg);
+        let msg = std::ffi::CStr::from_ptr(msg).to_string_lossy();
+        cb(Binding::from_raw(level), msg.as_ref());
     });
 }

--- a/src/tracing.rs
+++ b/src/tracing.rs
@@ -57,12 +57,9 @@ impl Binding for TraceLevel {
     }
 }
 
+//TODO: pass raw &[u8] and leave conversion to consumer (breaking API)
 /// Callback type used to pass tracing events to the subscriber.
 /// see `trace_set` to register a scubscriber.
-///
-/// Note:
-///     libgit2 might pass non-utf8 strings therefore we
-///     pass the message as a byte slice
 pub type TracingCb = fn(TraceLevel, &str);
 
 static CALLBACK: AtomicUsize = AtomicUsize::new(0);

--- a/src/tracing.rs
+++ b/src/tracing.rs
@@ -57,7 +57,7 @@ impl Binding for TraceLevel {
     }
 }
 
-pub type TracingCb = fn(TraceLevel, &str);
+pub type TracingCb = fn(TraceLevel, &[u8]);
 
 static CALLBACK: AtomicUsize = AtomicUsize::new(0);
 
@@ -76,7 +76,10 @@ extern "C" fn tracing_cb_c(level: raw::git_trace_level_t, msg: *const c_char) {
     let cb = CALLBACK.load(Ordering::SeqCst);
     panic::wrap(|| unsafe {
         let cb: TracingCb = std::mem::transmute(cb);
-        let msg = std::ffi::CStr::from_ptr(msg).to_string_lossy();
-        cb(Binding::from_raw(level), msg.as_ref());
+
+        if !msg.is_null() {
+            let msg = std::ffi::CStr::from_ptr(msg).to_bytes();
+            cb(Binding::from_raw(level), msg);
+        }
     });
 }

--- a/src/tracing.rs
+++ b/src/tracing.rs
@@ -63,7 +63,7 @@ impl Binding for TraceLevel {
 /// Note:
 ///     libgit2 might pass non-utf8 strings therefore we
 ///     pass the message as a byte slice
-pub type TracingCb = fn(TraceLevel, &[u8]);
+pub type TracingCb = fn(TraceLevel, &str);
 
 static CALLBACK: AtomicUsize = AtomicUsize::new(0);
 
@@ -82,10 +82,7 @@ extern "C" fn tracing_cb_c(level: raw::git_trace_level_t, msg: *const c_char) {
     let cb = CALLBACK.load(Ordering::SeqCst);
     panic::wrap(|| unsafe {
         let cb: TracingCb = std::mem::transmute(cb);
-
-        if !msg.is_null() {
-            let msg = std::ffi::CStr::from_ptr(msg).to_bytes();
-            cb(Binding::from_raw(level), msg);
-        }
+        let msg = std::ffi::CStr::from_ptr(msg).to_string_lossy();
+        cb(Binding::from_raw(level), msg.as_ref());
     });
 }

